lib/maxmind/geoip2/client.rb

```diff
@@ -0,0 +1,315 @@
+    end
+
+    private
+
+    def response_for(endpoint, model_class, ip_address)
+      record = get(endpoint, ip_address)
+
+      model_class.new(record, @locales)
+    end
+
+    # rubocop:disable Metrics/CyclomaticComplexity
+    # rubocop:disable Metrics/PerceivedComplexity
+    def get(endpoint, ip_address)
+      url = 'https://' + @host + '/geoip/v2.1/' + endpoint + '/' + ip_address
+
+      headers = HTTP.basic_auth(user: @account_id, pass: @license_key)
+                    .headers(
+                      accept: 'application/json',
+                      user_agent: 'MaxMind-GeoIP2-ruby',
+                    )
+      timeout = @timeout > 0 ? headers.timeout(@timeout) : headers
+
+      proxy = timeout
+      if @proxy_address != ''
+        opts = {}
+        opts[:proxy_port] = @proxy_port if @proxy_port != 0
+        opts[:proxy_username] = @proxy_username if @proxy_username != ''
+        opts[:proxy_password] = @proxy_password if @proxy_password != ''
+        proxy = timeout.via(@proxy_address, opts)
+      end
+
+      response = proxy.get(url)
+
+      body = response.to_s
+      is_json = response.headers[:content_type]&.include?('json')
+
+      if response.status.client_error?
+        return handle_client_error(endpoint, response.code, body, is_json)
+      end
+
+      if response.status.server_error?
+        raise HTTPError,
+              "Received server error response (#{response.code}) for #{endpoint} with body #{body}"
+      end
+
+      if response.code != 200
+        raise HTTPError,
+              "Received unexpected response (#{response.code}) for #{endpoint} with body #{body}"
+      end
+
+      handle_success(endpoint, body, is_json)
+    end
+    # rubocop:enable Metrics/CyclomaticComplexity
+    # rubocop:enable Metrics/PerceivedComplexity
+
+    # rubocop:disable Metrics/CyclomaticComplexity
+    def handle_client_error(endpoint, status, body, is_json)
+      if !is_json
+        raise HTTPError,
+              "Received client error response (#{status}) for #{endpoint} but it is not JSON: #{body}"
+      end
+
+      error = JSON.parse(body)
+
+      if !error.key?('code') || !error.key?('error')
+        raise HTTPError,
+              "Received client error response (#{status}) that is JSON but does not specify code or error keys: #{body}"
+      end
+
+      case error['code']
+      when 'IP_ADDRESS_INVALID', 'IP_ADDRESS_REQUIRED'
+        raise AddressInvalidError, error['error']
+      when 'IP_ADDRESS_NOT_FOUND'
+        raise AddressNotFoundError, error['error']
+      when 'IP_ADDRESS_RESERVED'
+        raise AddressReservedError, error['error']
+      when 'ACCOUNT_ID_REQUIRED',
+            'ACCOUNT_ID_UNKNOWN',
+            'AUTHORIZATION_INVALID',
+            'LICENSE_KEY_REQUIRED',
+            'USER_ID_REQUIRED',
+            'USER_ID_UNKNOWN'
+        raise AuthenticationError, error['error']
+      when 'OUT_OF_QUERIES', 'INSUFFICIENT_FUNDS'
```

Comment:
Similarly, I don't think we ever return `OUT_OF_QUERIES`, although I am less sure on that one. It looks like the web service docs might be out of date.